<body>
    <section style="border:1px solid RoyalBlue;">
        <section style="background-color:White; font-family:Georgia;text-align:center">
            <h1 style="color:RoyalBlue">Social Media Mining</h1>
            <h2 style="color:RoyalBlue">Dr.	Francesca	Spezzano</h1>
            <h2 style="font-family:Courier; text-align:center;">CS-539</h2>
            <br>
            <h2 style="font-family:Garamond;">Gerardo Caracas Uribe<br>Homework 1</h2>
            <hr/>
        </section>
    </section>
</body>

<!--Progress:<progress value="0" max="100"></progress>-->
<body>
    <section style="background-color:White; font-family:Georgia;text-align:center">
        <h3 style="font-family:Garamond;">Dataset Link<br></h3>
        <a href="http://snap.stanford.edu/data/soc-sign-bitcoin-otc.html">Bitcoin OTC trust weighted signed network</a>
        <hr/>
    </section>
</body>

<body>
    <section style="background-color:White; font-family:Georgia;text-align:center">
        <h3 style="font-family:Garamond;">Import the needed packages</h3>
        <hr/>
    </section>
</body>

In [2]:
import pandas as pd
import numpy as np
import networkx as nx
from datascience import *

<body>
    <section style="background-color:White; font-family:Georgia;text-align:center">
        <h3 style="font-family:Garamond;">Import and pre-process data</h3>
        <hr/>
    </section>
</body>

<body>
    <section style="background-color:White; font-family:Georgia;">
        <p style="font-family:Garamond;"><font size="4">Read the dataset file and convert it into a pandas dataframe</p>
    </section>
</body>

In [3]:
data = pd.read_csv('soc-sign-bitcoinotc.csv',names=['source','target','rating','time'], delimiter=',', encoding='ISO-8859-1')

<body>
    <section style="background-color:White; font-family:Georgia;">
        <p style="font-family:Garamond;"><font size="4">Data Sample: Showing the first 10 elements of the main dataframe containing all the edges data</p>
    </section>
</body>

In [4]:
data[0:4]

source  target  rating          time
0       6       2       4  1.289242e+09
1       6       5       2  1.289242e+09
2       1      15       1  1.289243e+09
3       4       3       7  1.289245e+09

<body>
    <section style="background-color:White; font-family:Georgia;text-align:center">
        <h3 style="font-family:Garamond;">Create Graph</h3>
        <hr/>
    </section>
</body>

In [ ]:
G = nx.DiGraph()
for line in data.iterrows():
    G.add_edge(line[1].source, line[1].target, weight=line[1].rating, time=line[1].time)

<body>
    <section style="background-color:White; font-family:Georgia;text-align:center">
        <h2 style="font-family:Garamond;">Excersise 2, section 4</h2>
        <hr/>
    </section>
</body>

<body>
    <section style="background-color:White; font-family:Georgia;text-align:center">
        <h2 style="font-family:Garamond;color:solid #A93226">Result #1/9</h2>
        <h3 style="font-family:Garamond;">Calculate the number of nodes</h3>
        <hr/>
    </section>
</body>

In [107]:
print("Number of nodes = ",G.number_of_nodes())

Number of nodes =  5881


<body>
    <section style="background-color:White; font-family:Georgia;text-align:center">
        <h2 style="font-family:Garamond;color:solid #A93226">Result #2/9</h2>
        <h3 style="font-family:Garamond;">Calculate the number of edges</h3>
        <hr/>
    </section>
</body>

In [108]:
print("Number of edges = ",G.number_of_edges())
globalNumEdges = G.number_of_edges()

Number of edges =  35592


<body>
    <section style="background-color:White; font-family:Georgia;">
        <p style="font-family:Garamond;"><font size="4">To traverse G using depth-first traversal of its edges, we need to extract a list of nodes and a list of edges</p>
    </section>
</body>

In [109]:
edges=list(G.edges())
nodes=list(G.nodes())

<body>
    <section style="background-color:White; font-family:Georgia;">
        <p style="font-family:Garamond;"><font size="4">Now let's get a list of the graph traversal</p>
    </section>
</body>

In [110]:
traverseList = list(nx.edge_dfs(nx.DiGraph(edges), nodes))

<body>
    <section style="background-color:White; font-family:Georgia;">
        <p style="font-family:Garamond;"><font size="4">Now that we have the graph traversal data, let's iterate and analyze the weights</p>
    </section>
</body>

In [11]:
traverseSet = set(traverseList)
positiveW = []
negativeW = []
mainEdgeTable = Table()
for source, target in traverseSet:
    if G.edges[source, target]['weight'] < 0: 
        negativeW.append(G.edges[source, target]['weight'])
    else:
        positiveW.append(G.edges[source, target]['weight'])

<body>
    <section style="background-color:White; font-family:Georgia;text-align:center">
        <br>
        <h2 style="font-family:Garamond;color:solid #A93226">Result #3/9</h2>
        <h3 style="font-family:Garamond;">Percentage of negative edges</h3>
        <br>
        <hr/>
    </section>
</body>

In [12]:
print(" Negative edges percent = ",(len(negativeW) * 100 )/ globalNumEdges,"%")

 Negative edges percent =  10.010676556529557 %


<body>
    <section style="background-color:White; font-family:Georgia;text-align:center">
        <h2 style="font-family:Garamond;">Excersise 2, section 5</h2>
        <hr/>
    </section>
</body>

<body>
    <section style="background-color:White; font-family:Georgia;text-align:center">
        <h3 style="font-family:Garamond;">Convert data into a datascience's package Tabe()</h3>
        <hr/>
    </section>
</body>

In [14]:
mainEdgeTable = Table().with_columns('source',list(data['source']), 
                                    'target', list(data['target']),
                                    'rating', list(data['rating']),
                                    'time', list(data['time']))

<body>
    <section style="background-color:White; font-family:Georgia;">
        <p style="font-family:Garamond;"><font size="4">Data Sample: Showing the first few elements of the new table</p>
    </section>
</body>

In [112]:
mainEdgeTable

source | target | rating | time
1128   | 13     | 2      | 1.45368e+09
13     | 1128   | 1      | 1.45368e+09
2731   | 4897   | 5      | 1.45368e+09
2731   | 3901   | 5      | 1.45368e+09
4499   | 1810   | 1      | 1.45361e+09
1810   | 4499   | 2      | 1.45361e+09
1953   | 5655   | -10    | 1.45356e+09
13     | 4608   | 2      | 1.45332e+09
2045   | 4608   | 2      | 1.45332e+09
4608   | 2045   | 2      | 1.45332e+09
... (35582 rows omitted)

<body>
    <section style="background-color:White; font-family:Georgia;">
        <p style="font-family:Garamond;"><font size="4">This new table allow us to do many operations, in this case we will sort it, and then we can take the max and min dates</p>
    </section>
</body>

In [114]:
mainEdgeTable = mainEdgeTable.sort('time',descending=True)

<body>
    <section style="background-color:White; font-family:Georgia;">
        <p style="font-family:Garamond;"><font size="4">Now we just take the last and first element</p>
    </section>
</body>

In [116]:
maxtime = mainEdgeTable.take(0)
mintime = mainEdgeTable.take(mainEdgeTable.num_rows-1)

<body>
    <section style="background-color:White; font-family:Georgia;text-align:center">
        <br>
        <h2 style="font-family:Garamond;color:solid #A93226">Result #4/9</h2>
        <h3 style="font-family:Garamond;">Compute the time window of network creation, i.e. the minimum and maximum edge creation timestamps</h3>
        <br>
        <br>
    </section>
</body>

In [126]:
print("The time window is from time = ",mintime['time']," and time = ",maxtime['time'])

The time window is from time =  [1.28924191e+09]  and time =  [1.45368432e+09]


<body>
    <section style="background-color:White; font-family:Georgia;text-align:center">
        <h2 style="font-family:Garamond;">Excersise 2, section 6</h2>
        <hr/>
    </section>
</body>

<body>
    <section style="background-color:White; font-family:Georgia;text-align:center">
        <br>
        <h3 style="font-family:Garamond;color:solid #A93226">Note: This excercise requires us to convert the graph to a directed one, however this graph is already directed, thus we won't do this task</h3>
        <br>
        <br>
    </section>
</body>

<body>
    <section style="background-color:White; font-family:Georgia;text-align:center">
        <h3 style="font-family:Garamond;">Let's calculate the Graph's diameter now</h3>
        <hr/>
    </section>
</body>

<body>
    <section style="background-color:White; font-family:Georgia;">
        <p style="font-family:Garamond;"><font size="4">We need first see how many strongly components we have and weak ones.</p><br>
        <pstyle="font-family:Garamond;"><font size="4">This is required because we can not calculate the diameter of a directed weakly connected component</p>
    </section>
</body>

In [122]:
print("The number of strongly connected components = ",nx.number_strongly_connected_components(G))

The number of strongly connected components =  1144


In [123]:
print("Number of weakly connected components = ",nx.number_weakly_connected_components(G))

Number of weakly connected components =  4


<body>
    <section style="background-color:White; font-family:Georgia;">
        <p style="font-family:Garamond;"><font size="4">We know now that we have some weakly connected components. Therefore we will now get a list of the strongly connected compoents only, in the form of sub graphs</p>
    </section>
</body>

In [124]:
stConnectedC = list(nx.strongly_connected_component_subgraphs(G, copy=False))

<body>
    <section style="background-color:White; font-family:Georgia;">
        <p style="font-family:Garamond;"><font size="4">This is a very expensive process, which is calculate the diameter of each one of the subgraphs (Which all of them are strongly connected) and get the maximum diameter from them</p>
    </section>
</body>

In [22]:
maxD = 0
for graph in stConnectedC:
    # Trying to save some processing by only processing components which have more than 1 node
    if graph.number_of_nodes() > 1:
        diam = nx.diameter(graph)
        print("nodes = ", graph.number_of_nodes(), "diameter = ", diam)
        if diam > maxD:
            maxD = diam
print("For the string connected components, the diameter is ", maxD)            

nodes =  2 diameter =  1
nodes =  2 diameter =  1
nodes =  2 diameter =  1
nodes =  3 diameter =  2
nodes =  2 diameter =  1
nodes =  2 diameter =  1
nodes =  2 diameter =  1
nodes =  2 diameter =  1
nodes =  2 diameter =  1
nodes =  6 diameter =  2
nodes =  3 diameter =  2
nodes =  2 diameter =  1
nodes =  2 diameter =  1
nodes =  2 diameter =  1
nodes =  2 diameter =  1
nodes =  2 diameter =  1
nodes =  2 diameter =  1
nodes =  3 diameter =  2
nodes =  4709 diameter =  11
nodes =  2 diameter =  1
nodes =  2 diameter =  1
nodes =  2 diameter =  1
nodes =  2 diameter =  1
For the string connected components, the diameter is  11


<body>
    <section style="background-color:White; font-family:Georgia;">
        <p style="font-family:Garamond;"><font size="4">We know now that we have some weakly connected components. Therefore we will now get a list of the strongly connected compoents only, in the form of sub graphs</p>
    </section>
</body>

<body>
    <section style="background-color:White; font-family:Georgia;text-align:center">
        <br>
        <h2 style="font-family:Garamond;color:solid #A93226">Result #5/9</h2>
        <h3 style="font-family:Garamond;">Compute the diameter of the network</h3>
        <br>
        <br>
    </section>
</body>

In [23]:
print("The Diameter of the network is ",maxD)

The Diameter of the network is  11


## Notes on Diameter<br>
This graph, has 1144 strongly connected components, and 4 weakly connected components.<br>
The result given as diameter = 11, corresponds to the highest diameter from all the strongly connected components; however we can not calculate the diameter on weakly connected components due to this definition:<br>
*Definition 1.5.17. Let G = (V, E) be a graph (directed or not). The diameter of G, denoted
diam(G), is the maximum distance d(u, v) for u, v ∈ V .
The diameter is finite if and only if G is connected and undirected, or G is strongly connected
and directed.*<br>
source: "Graph Theory and Social Networks
Spring 2014 Notes" by Kimball Martin
April 30, 2014, University of Oklahoma, Department of mathematics. http://www2.math.ou.edu/~kmartin/graphs/graphs.pdf, page 46<br>

<body>
    <section style="background-color:White; font-family:Georgia;text-align:center">
        <h3 style="font-family:Garamond;">Calculate now the maximum and minimum degree, in-degree and out-degree of the network</h3>
        <hr/>
    </section>
</body>

<body>
    <section style="background-color:White; font-family:Georgia;text-align:center">
        <br>
        <h2 style="font-family:Garamond;color:solid #A93226">Result #6/9</h2>
        <h3 style="font-family:Garamond;">Max and Min degree, in-degree and out-degreek</h3>
        <br>
        <br>
    </section>
</body>

In [100]:
print("The minimum in_degree in graph is:")
min(dict(set(G.in_degree(G.nodes()))).values())

The minimum in_degree in graph is:


0

In [101]:
print("The maximum in_degree in graph is")
max(dict(set(G.in_degree(G.nodes()))).values())

The maximum in_degree in graph is


535

In [102]:
print("The minimum out_degree in graph is:")
min(dict(set(G.out_degree(G.nodes()))).values())

The minimum out_degree in graph is:


0

In [104]:
print("The maximum out_degree in graph is:")
max(dict(set(G.out_degree(G.nodes()))).values())

The maximum out_degree in graph is:


763

In [105]:
print("The minimum degree in graph is:")
min(dict(set(G.degree(G.nodes()))).values())

The minimum degree in graph is:


1

In [106]:
print("The maximum degree in graph is:")
max(dict(set(G.degree(G.nodes()))).values())

The maximum degree in graph is:


1298

<body>
    <section style="background-color:White; font-family:Georgia;text-align:center">
        <h2 style="font-family:Garamond;">Excersise 2, section 7</h2>
        <hr/>
    </section>
</body>

<body>
    <section style="background-color:White; font-family:Georgia;text-align:center">
        <br>
        <h2 style="font-family:Garamond;color:solid #A93226">Result #7/9</h2>
        <h3 style="font-family:Garamond;">Whether or not the graph is connected</h3>
        <br>
        <br>
    </section>
</body>

In [131]:
# In a directed graph we can not ask the question "connected" but instead strongly or weakly connected
nx.is_strongly_connected(G)

False

<body>
    <section style="background-color:White; font-family:Georgia;text-align:center">
        <br>
        <h2 style="font-family:Garamond;color:solid #A93226">Result #8/9</h2>
        <h3 style="font-family:Garamond;">How many Strongly connected components does the graph have?</h3>
        <br>
        <br>
    </section>
</body>

In [133]:
print("The number of strongly connected components = ",nx.number_strongly_connected_components(G))

The number of strongly connected components =  1144


<body>
    <section style="background-color:White; font-family:Georgia;text-align:center">
        <br>
        <h2 style="font-family:Garamond;color:solid #A93226">Result #9/9</h2>
        <h3 style="font-family:Garamond;">How many Weakly connected components does the graph have?</h3>
        <br>
        <br>
    </section>
</body>

In [135]:
print("Number of weakly connected components = ",nx.number_weakly_connected_components(G))

Number of weakly connected components =  4
